### ISSUE TITLE
how to define a torch function within python?

### ISSUE THREAD

**Issue by TuBui** [imodpasteur/lutorpy#33]
I want to use the 'optim' package for the optimisation process. I follow the example [here](https://github.com/torch/optim/blob/master/doc/intro.md) to define the feval() function within python:

```
require('optim')
for epoch in range(50):
   function feval(params)
      gradParams:zero()
      local outputs = model:forward(batchInputs)
      local loss = criterion:forward(outputs, batchLabels)
      local dloss_doutputs = criterion:backward(outputs, batchLabels)
      model:backward(batchInputs, dloss_doutputs)
      return loss, gradParams
   end
   optim.sgd(feval, params, optimState)
end 
```

But python claims it is a invalid syntax. I cannot put the function definition in a separate .lua file then import it to python because it nees to access many shared variables such as gradParams, model, criterion,...

Is there a way to overcome it?

**Comment by oeway** [comment]
You can always define a trainer class in a lua file, then you can use it as a regular object in python.


For example, you can define a trainer module like this:
```lua
require 'nn'
require 'optim'
require 'nngraph'

if Trainer == nil then
    local Trainer = torch.class('Trainer')
end

function Trainer:__init(options)
    options = options or {}
    self.coefL1 = options.coefL1 or 0
    self.coefL2 = options.coefL2 or 0
    self.cudaMode = options.cudaMode or false
    self.criterion = options.criterion or nn.MSECriterion()
    self.optimizer = options.optimizer or 'sgd'
    self:setOptimizer(self.optimizer)
    self.epochCount = 0
    if options.model then
        self:setModel(options.model)
    end
end

function Trainer:setModel(model)
    self.model = model
    if model then
        self:setMode(self.cudaMode)
        self.parameters,self.gradParameters = self.model:getParameters()
        return true
    else
        error()
    end
end


function Trainer:setMode(cudaMode)
    if not self.model then
       error()
    end
    
    if cudaMode then
        require 'cudnn'
        self.model:cuda()
        cudnn.convert(self.model, cudnn)
        self.criterion:cuda()
    else
        self.model:float()
        self.criterion:float()
    end
    self.cudaMode = cudaMode
    self.model.cudaMode = cudaMode
    return cudaMode
end

function Trainer:setOptimizer(optimizer)
    if optimizer == 'sgd' then
        self.optimMethod = optim.sgd
        self.optimizer_params = {
           learningRate = 1e-2,
           learningRateDecay = 1e-4,
           weightDecay = 1e-4,
           momentum = 0.9,
           dampening = 0,
           nesterov = true
        }
    elseif optimizer == 'adadelta' then
        self.optimMethod = optim.adadelta
        self.optimizer_params = {
           rho = 0.9,
           eps = 1e-6,
        }
    elseif optimizer == 'rmsprop' then
        self.optimMethod = optim.rmsprop
        self.optimizer_params = {
           learningRate = 1e-2,
           alpha = 0.99,
           epsilon = 1e-8,
        }
    elseif optimizer == 'adagrad' then
        self.optimMethod = optim.adagrad
        self.optimizer_params = {
           learningRate = 1e-3,
           learningRateDecay = 0,
           weightDecay = 0,
        }
    else
        error('unsupported optimizer: ' .. optimizer)
    end 
end

function Trainer:batchTrain(trainData, batchSize, batchCallback)
    if not self.model then
        error()
    end
    batchSize = batchSize or 1
    -- shuffle at each epoch
    local shuffle = torch.randperm(trainData:size()):long()
    -- do one epoch
    -- io.write('==> online epoch # ' .. self.epochCount+1 .. ' [batchSize = ' .. batchSize .. ']')
    local fs = nil
    local epochLoss = 0
    local batchCount = 0
    for t = 1, trainData:size(), batchSize do
        -- disp progress
        -- xlua.progress(t, trainData:size())
        local time = sys.clock()
        local feval
       
        -- create mini batch
        local inputs = trainData.x:index(1,shuffle[{{t,t+batchSize-1}}])
        local targets = trainData.y:index(1,shuffle[{{t,t+batchSize-1}}])
        if self.cudaMode then
            inputs = inputs:cuda()
            targets = targets:cuda()
        end
        feval = function(x)
            -- get new parameters
            if x ~= self.parameters then
                self.parameters:copy(x)
            end
            -- reset gradients
            self.gradParameters:zero()
            local outputs = self.model:forward(inputs)
            local f = self.criterion:forward(outputs, targets)
            -- estimate df/dW
            local df_do = self.criterion:backward(outputs, targets)
            self.model:backward(inputs, df_do)
            -- penalties (L1 and L2):
            if self.coefL1 ~= 0 or self.coefL2 ~= 0 then
                -- locals:
                local norm,sign= torch.norm,torch.sign
                -- Loss:
                f = f + opt.coefL1 * norm(self.parameters,1)
                f = f + opt.coefL2 * norm(self.parameters,2)^2/2
                -- Gradients:
                self.gradParameters:add( sign(self.parameters):mul(opt.coefL1) + self.parameters:clone():mul(opt.coefL2) )
            end
            -- normalize gradients and f(X)
            self.gradParameters:div(inputs:size(1))
            -- return f and df/dX
            return f,self.gradParameters
        end

        self.model:training()
        _, fs = self.optimMethod(feval, self.parameters, self.optimizer_params)
        self.model:evaluate()
        
        local loss = fs[1]/batchSize
        epochLoss = epochLoss + loss
        batchCount = batchCount + 1
        -- time taken
        time = sys.clock() - time
        if batchCallback then
            batchCallback(batchCount, loss, time)
        end
    end
    -- io.write('  loss: ' .. loss .. '\n')
    self.epochCount = self.epochCount + 1
    epochLoss = epochLoss/batchCount
    return epochLoss
end

function Trainer:train(trainDataset, epochNum, batchSize, epochCallback, batchCallback)
    if not self.model then
        error()
    end
    epochNum = epochNum or 1
    batchSize = batchSize or 1
    print(string.format('==> Training on %d samples for %d epochs (batch size: %d) <==',trainDataset.size(), epochNum, batchSize))
    local time = sys.clock()
    for i = 1,epochNum do
        local etime = sys.clock()
        local loss, time = self:batchTrain(trainDataset, batchSize, batchCallback)
        -- time taken
        etime = sys.clock() - etime
        if epochCallback then
            epochCallback(i, loss, etime)
        end
    end
    time = sys.clock() - time
    print(string.format('==> Trained on %d samples for %d epochs. (total time: %.2fs) <==',trainDataset.size(), epochNum, time))
end

return Trainer
```
Save it as file `trainer.lua` into a folder named `mymodules` for example, after, you will be able to access by `require("mymodules.trainer")`.

Then, in lutorpy, you can have something like this:
```python
import lutorpy
require("nn")
require("mymodules.trainer")

model = nn.Sequential()
model.add(...)

t = Trainer(model=model, cudaMode=False)
t._train(trainDataset, epochNum, batchSize, epochCallback, batchCallback)
```


The code is modified from my previous torch code but I haven't test it, you may need some minor change in order to use it. But here I just want to show you how can we define a trainer in torch.

**Comment by TuBui** [comment]
@oeway thank you for your detailed code. Although my problem does not allow me to put everything in a neat & tidy trainer.lua (e.g. I need to manipulate every training minibatch in python -- the main reason I come to lutorpy ^_^), I see the way you deal with the problem now. It is of great helps.

One piece of your code troubles me though, what is epochCallback and batchCallBack? From the way it is passed into the `train()` function, I guess they are just boolean variables to set verbosity. However the line `epochCallback(i, loss, etime) ` makes me think it is a function. How possible is it?

**Comment by oeway** [comment]
@TuBui Welcome, glad that it's of help.
So for your question, those callback are functions, you see some if statement with those functions because if they are not passed, then the value will be nil. So those callbacks are optional arguments for the function. Which I used them for print information, save weights etc.

**Comment by oeway** [comment]
By the way, you know, you can also manipulate the training through those callbacks, for example if you pass the model into callbacks, then you can still manipulate the model parameters inside callbacks. And callbacks are defined in python. I guess that will be more close to what you want.
